### ISSUE TITLE
`unicorn master (old)` hanging around

### ISSUE THREAD

**Issue by docwhat** [sosedoff/capistrano-unicorn#8]
When doing a `cap deploy` the old master still hangs around.

See [SIGNALS](http://unicorn.bogomips.org/SIGNALS.html) in the unicorn docs.  There should be a task or something to kill off the old master or roll back to old master and kill the new one.

Ciao!

**Comment by sosedoff** [comment]
Hey,

I dont know the real issue behind it even though it happens to my apps for unknown reason. Like 95% of time it works properly, but some time the master worker is hanging and there only one way to kill it, manually. 

Although, regarding the signals, isnt the `unicorn:reload` doing it ?

**Comment by docwhat** [comment]
I'm going to fork your project and see if I can figure out what's going on.  I'm certainly no unicorn expert, but there are two of us here that need it and between the two of us and the docs, we ought to be able to make mess`^H^H^H^H` short work of it. ;-)

**Comment by angelim** [comment]
I was having the same issue until using before_fork and after_fork to signal the old process to quit. The example provided in https://github.com/sosedoff/capistrano-unicorn/blob/master/examples/rails3.rb was really helpful.

**Comment by docwhat** [comment]
Oh...well, part of the problem I was having was I didn't use that because the file said `rails3.rb` and I was working on Sinatra.

The for stuff still doesn't look right. I still don't see the WINCH in there to switch to the new process.  I'm also not sure disconnecting is required either.

**Comment by angelim** [comment]
The last worker for the new master signals to kill the old master. The commented example configuration in the unicorn repository should explain this a lot better. https://github.com/defunkt/unicorn/blob/master/examples/unicorn.conf.rb
I hope it helps. It's surely working for me now.

**Comment by docwhat** [comment]
Is there any reason not to move it into `capistrano-unicorn` directly? It'd make configuring unicorn trivial.

**Comment by docwhat** [comment]
My incorrect assumption when I picked up `capistrano-unicorn` was that it would configure everything in unicorn that capistrano already knows about:
- `pid` file location
- log location (`stdout`, `stderr`)
- `working_directory`
- and would reliably start, stop and reload unicorn without me having to think much about it.

**Comment by exviva** [comment]
@docwhat why is that an incorrect assumption? Isn't that doable?

BTW I'm experiencing similar issues, and it looks like my nginx is still proxying to the old master processes.

**Comment by docwhat** [comment]
It's doable but it isn't what capistrano-unicorn does.  You have to do it yourself in the unicorn configuration file.

**Comment by exviva** [comment]
Now I see what you mean, thanks.

**Comment by masterkain** [comment]
I have the same problem, using rainbows, the old master sticks around.

**Comment by masterkain** [comment]
I used a similar code to resolve the 'issue':

```
  old_pid = "#{server.config[:pid]}.oldbin"
  if old_pid != server.pid
    begin
      sig = (worker.nr + 1) >= server.worker_processes ? :QUIT : :TTOU
      Process.kill(sig, File.read(old_pid).to_i)
    rescue Errno::ENOENT, Errno::ESRCH
    end
  end
```

**Comment by sosedoff** [comment]
This will be fixed in 0.2 version, which is in works right now.

**Comment by sosedoff** [comment]
Few explanations about configs:
1. PID file location is not required since capistrano-unicorn sets it. Most people wont change it and it is also a configurable option.
2. `stderr_path` and `stdout_path` are defaulted to /dev/null according to unicorn documentation. Unicorn does not have a CLI agrument that allows you to set this parameter so the easiest way to set path is from unicorn config file.  

Regarding the handing process issue - any of you use application preloading ? If so, there is a difference in signals and handling. Best option for reloading (not restarting) the app is to send HUP signal to the master process, which will reload config and workers. If you use preloading the code will not be updated, so you'd need to issue USR2 + QUIT signals for master process.
